Context-dependent SmtCore

AleksandarZeljic opened the pull request:
This PR introduces a class with context-dependent implementation of the SmtCore. 

The CDSmtCore class implementation relies on:
- Access to the unique context object from which all the context-dependent structures are obtained.
- PiecewiseLinearConstraint class stores its search state in a context-dependent manner (already implemented) and 
modifying it using nextFeasibleCase()/markInfeasible() methods.
- Using BoundManager class to store bounds in a context-dependent manner (to be integrated in a future patch).

The CDSmtCore distinguishes between: **decisions** and **implications**. 
Decision is a case of PiecewiseLinearConstraint (represented using PhaseStatus enum), which is being asserted on the trail. The fact that it is a decision implies that the PiecewiseLinearConstraint has at least one other feasible case.
Decisions represent nodes/search-states in the search-space tree.

Implication is a case of PiecewiseLinearConstraint (represented using PhaseStatus enum), which is being asserted on the trail. Implication represents the last feasible case of the PiecewiseLinearConstraint.

_trail is a chronological list of cases of PiecewiseLinearConstraints asserted to hold (represented using TrailEntry) - both decisions and implications.
_decisions is a chronological list of decisions stored on the trail. 
_trail and _decisions are both context dependent and will synchronize in unison with the context object.

When we split on cases of a PiecewiseLinearConstraint a decision is performed.
When a case of PiecewiseLinearConstraint is fixed via bound propagation an implication is performed.
When a search state is found to be infeasible, CDSmtCore backtracks lazily to the last decision and continues the search.

Context management is done automatically when a new decision is pushed and a decision level is popped.
Popping a decision will cause all related context dependent data structures to backtrack in sync with the _trail.
This will effectively backtrack the entire search state via PiecewiseLinearConstraints and boundManager. 
The only exception is the state of basic/non-basic variables in the tableau, which may need to be recalculated. 
However, no additional memory overhead is incurred in this process.

Next steps: Integration of the centralized context object, proper initialization of all context dependent structures and integrating them all through via the Engine and CDSmtCore class.

guykatzz commented on src/engine/Engine.h:
```diff
@@ -126,6 +126,11 @@ class Engine : public IEngine, public SignalHandler::Signalable
+    /*
+      Recompute status of basic variables
+    */
+    void recomputeBasicStatus() { _tableau->computeBasicStatus(); };
```
This doesn't seem like a method that should appear in the Engine.
Perhaps we should only expose a Tableau method called "contextPopHook", or something like that, that will get invoked after a pop?

AleksandarZeljic replied:
Great point, done!

guykatzz commented on src/engine/TrailEntry.h:
```diff
@@ -0,0 +1,61 @@
+      return _pwlConstraint->isFeasible();
+  }
+
+ TrailEntry(PiecewiseLinearConstraint * pwlc, PhaseStatus phase )
```
space after (

guykatzz commented on src/engine/CDSmtCore.h:
```diff
@@ -0,0 +1,270 @@
+/*! \file CDSmtCore.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Guy Katz, AleksandarZeljic,  Parth Shah
```
remove space before Parth

guykatzz commented on src/engine/CDSmtCore.h:
```diff
@@ -0,0 +1,270 @@
+ ** and _boundManager. The only exception is the state of basic/non-basic
+ ** variables in the tableau, which may need to be recalculated. However, no
+ ** additional memory overhead is incurred in this process.
+ **
```
These comments are great, well done!

guykatzz commented on src/engine/CDSmtCore.h:
```diff
@@ -0,0 +1,270 @@
+
+    /*
+        Pushes trail entry onto trail, handles decision book-keeping and
+        Update bounds and add equations to the engine.
```
Update --> update

guykatzz commented on src/engine/CDSmtCore.h:
```diff
@@ -0,0 +1,270 @@
+    void decideSplit( PiecewiseLinearConstraint *constraint );
+
+    /*
+       Backtracks fully explored decisions and stores the last (feasible decision
```
remove (

guykatzz commented on src/engine/CDSmtCore.h:
```diff
@@ -0,0 +1,270 @@
+      Split when some relu has been violated for this many times
+    */
+    unsigned _constraintViolationThreshold;
+
```
remove empty line

guykatzz commented on src/engine/CDSmtCore.cpp:
```diff
@@ -0,0 +1,432 @@
+/*! \file CDSmtCore.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Guy Katz, Aleksandar Zeljic,  Parth Shah
```
Remove space before Parth

guykatzz commented on src/engine/CDSmtCore.cpp:
```diff
@@ -0,0 +1,432 @@
+    decideSplit( _constraintForSplitting );
+}
+
+void CDSmtCore::decideSplit( PiecewiseLinearConstraint * constraint )
```
remove space after *

guykatzz reviewed (commented):
Hi Alex,
Looks very good. Most of my comments are cosmetic.

At various points you mention "lazily backtrack...". Where is this laziness expressed?

AleksandarZeljic commented:
Hi Guy, thanks for the comments, I'll fix them! 

All the context-dependent structures are inherently lazy. Backtracking updates the end pointer of the data structure to an earlier position - which effectively forgets everything beyond that point in 0(1) time. I can add a clarification to the comments.

guykatzz commented:
Okay. So laziness here just means that the actual freeing of the memory
happens at some later time?

On Wed, 10 Nov 2021 at 23:18, AleksandarZeljic ***@***.***>
wrote:

> Hi Guy, thanks for the comments, I'll fix them!
>
> All the context-dependent structures are inherently lazy. Backtracking
> updates the end pointer of the data structure to an earlier position -
> which effectively forgets everything beyond that point in 0(1) time. I can
> add a clarification to the comments.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/NeuralNetworkVerification/Marabou/pull/487#issuecomment-965753458>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSTPXOW2HHQA7ZIAIHTEHDULLOSTANCNFSM5HPASEWQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

AleksandarZeljic commented:
Yes, the memory is acquired by Context and it is only freed once at the very end. All the context-dependent data structures live in the Context's memory and their synchronized management is achieved via pointer updates.

AleksandarZeljic commented:
The key point of laziness here is that pointer update does not care about the size of memory being backtracked - whether it is one list element or a thousand the effort is the same.

AleksandarZeljic commented:
I've addressed all comments, @guykatzz, you can take another look. 

I've clarified the comments further to address the laziness, let me know if there is anything else that stands out.
